File: solutions/zig/dnd-character/1/dnd_character.zig
const std = @import("std");
const rand = std.crypto.random;

pub fn modifier(score: i8) i8 {
    return @divFloor(score - 10, 2);
}

pub fn ability() i8 {
    return rand.intRangeAtMost(i8, 3, 18);
}

pub const Character = struct {
    strength: i8,
    dexterity: i8,
    constitution: i8,
    intelligence: i8,
    wisdom: i8,
    charisma: i8,
    hitpoints: i8,

    pub fn init() Character {
        var c = Character{
            .strength = ability(),
            .dexterity = ability(),
            .constitution = ability(),
            .intelligence = ability(),
            .wisdom = ability(),
            .charisma = ability(),
            .hitpoints = 0,
        };
        c.hitpoints = 10 + modifier(c.constitution);
        return c;
    }
};

File: solutions/clojure/dnd-character/1/src/dnd_character.clj
(ns dnd-character)

(defn score-modifier [score]
  (Math/floorDiv (- score 10) 2))

(defn rand-ability []
  (->> (repeatedly 4 #(inc (rand-int 6)))
       sort
       rest
       (reduce +)))

(defn rand-character []
  (let [constitution (rand-ability)]
    (assoc (zipmap [:strength :dexterity :intelligence :wisdom :charisma]
                   (repeatedly 5 rand-ability))
           :constitution constitution
           :hitpoints (+ 10 (score-modifier constitution)))))
